File: Tests/swift-sdk-objc-tests/IterableAPIResponseTests.m
//
//  IterableAPIResponseTests.m
//  Iterable-iOS-SDKTests
//
//  Created by Victor Babenko on 3/29/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "OHHTTPStubs.h"
#import "OHPathHelpers.h"

@import IterableSDK;

static CGFloat const IterableResponseExpectationTimeout = 1.0;

@interface IterableAPIResponseTests : XCTestCase

@end

@implementation IterableAPIResponseTests

- (void)setUp {
    [super setUp];
    [IterableAPI initializeWithApiKey:@""];
}

- (void)tearDown {
    [OHHTTPStubs removeAllStubs];
    [super tearDown];
}

- (NSDictionary *)defaultResponseHeaders {
    return @{@"Content-Type":@"application/json"};
}

- (void)stubAnyRequestReturningStatusCode:(int)statusCode data:(NSData *)data {
    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
        return YES;
    } withStubResponse:^OHHTTPStubsResponse * _Nonnull(NSURLRequest * _Nonnull request) {
        return [OHHTTPStubsResponse responseWithData:data statusCode:statusCode headers:self.defaultResponseHeaders];
    }];
}

- (void)stubAnyRequestReturningStatusCode:(int)statusCode json:(NSDictionary *)json {
    NSError *error = nil;
    NSData *jsonResponseData = [NSJSONSerialization dataWithJSONObject:json
                                                               options:0 error:&error];
    [self stubAnyRequestReturningStatusCode:statusCode data:jsonResponseData];
}

- (void)testResponseCode200 {
    NSDictionary *responseData = @{@"key":@"value"};
    [self stubAnyRequestReturningStatusCode:200 json:responseData];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onSuccess is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:^(NSDictionary * _Nonnull data) {
        [expectation fulfill];
        XCTAssert([data isEqualToDictionary:responseData]);
    } onFailure:nil];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testResponseCode200WithNoData {
    [self stubAnyRequestReturningStatusCode:200 data:nil];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssertEqualObjects(reason, @"No data received");
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testResponseCode200WithInvalidJson {
    [self stubAnyRequestReturningStatusCode:200 data:[@"{'''}}" dataUsingEncoding:NSUTF8StringEncoding]];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssert([reason containsString:@"Could not parse json"]);
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testResponseCode400WithoutMessage {
    [self stubAnyRequestReturningStatusCode:400 json:@{}];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssert([reason containsString:@"Invalid Request"]);
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testResponseCode400WithMessage {
    [self stubAnyRequestReturningStatusCode:400 json:@{@"msg":@"Test error"}];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssertEqualObjects(reason, @"Test error");
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testResponseCode401 {
    [self stubAnyRequestReturningStatusCode:401 json:@{}];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssertEqualObjects(reason, @"Invalid API Key");
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testResponseCode500 {
    [self stubAnyRequestReturningStatusCode:500 json:@{}];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssertEqualObjects(reason, @"Internal Server Error");
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testNon200ResponseCode {
    [self stubAnyRequestReturningStatusCode:302 json:@{}];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssert([reason containsString:@"Received non-200 response"]);
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testNoNetworkResponse {
    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
        return YES;
    } withStubResponse:^OHHTTPStubsResponse * _Nonnull(NSURLRequest * _Nonnull request) {
        NSError* notConnectedError = [NSError errorWithDomain:NSURLErrorDomain code:kCFURLErrorNotConnectedToInternet userInfo:nil];
        return [OHHTTPStubsResponse responseWithError:notConnectedError];
    }];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSURLRequest *request = [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssert([reason containsString:@"NSURLErrorDomain"]);
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

- (void)testNetworkTimeoutResponse {
    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
        return YES;
    } withStubResponse:^OHHTTPStubsResponse * _Nonnull(NSURLRequest * _Nonnull request) {
        NSError* notConnectedError = [NSError errorWithDomain:NSURLErrorDomain code:kCFURLErrorNotConnectedToInternet userInfo:nil];
        return [[OHHTTPStubsResponse responseWithData:nil statusCode:200 headers:nil] requestTime:0.0 responseTime:5.0];
    }];
    
    XCTestExpectation *expectation = [[XCTestExpectation alloc] initWithDescription:@"onFailure is called"];
    
    NSMutableURLRequest *request = (NSMutableURLRequest *) [[IterableAPI sharedInstance] createPostRequestForAction:@"" withArgs:@{}];
    request.timeoutInterval = 0.01;
    [[IterableAPI sharedInstance] sendRequest:request onSuccess:nil onFailure:^(NSString * _Nonnull reason, NSData * _Nullable data) {
        [expectation fulfill];
        XCTAssert([reason containsString:@"timed out"]);
    }];
    [self waitForExpectations:@[expectation] timeout:IterableResponseExpectationTimeout];
}

@end

File: tests/swift-sdk-swift-tests/IterableAPIResponseTests.swift
//
//  Created by Tapash Majumder on 9/5/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

import OHHTTPStubs
import XCTest

@testable import IterableSDK

class IterableAPIResponseTests: XCTestCase {
    private let apiKey = "zee_api_key"
    private let email = "user@example.com"
    private let authToken = "asdf"
    
    func testHeadersInGetRequest() {
        let iterableRequest = IterableRequest.get(GetRequest(path: "", args: ["var1": "value1"]))
        let urlRequest = createApiClient().convertToURLRequest(iterableRequest: iterableRequest)!
        
        verifyIterableHeaders(urlRequest)
    }
    
    func testHeadersInPostRequest() {
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: ["var1": "value1"], body: [:]))
        let urlRequest = createApiClient().convertToURLRequest(iterableRequest: iterableRequest)!
        
        verifyIterableHeaders(urlRequest)
    }
    
    func testAuthInHeader() {
        let apiClient = createApiClientWithAuthToken()
        
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: ["var1": "value1"], body: [:]))
        let urlRequest = apiClient.convertToURLRequest(iterableRequest: iterableRequest)!
        
        verifyIterableHeaders(urlRequest)
        verifyAuthTokenInHeader(urlRequest, authToken)
    }
    
    func testResponseCode200() {
        let xpectation = expectation(description: "response code 200")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient().send(iterableRequest: iterableRequest).onSuccess { _ in
            xpectation.fulfill()
        }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testResponseCode200WithNoData() {
        let xpectation = expectation(description: "no data")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 200, data: nil))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("no data"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testResponseCode200WithInvalidJson() {
        let xpectation = expectation(description: "invalid json")
        let data = "{'''}}".data(using: .utf8)!
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 200, data: data))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("could not parse json"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testResponseCode400WithoutMessage() { // 400 = bad reqeust
        let xpectation = expectation(description: "400 without message")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 400))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("invalid request"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testResponseCode400WithMessage() {
        let xpectation = expectation(description: "400 with message")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 400, json: ["msg": "Test error"]))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("test error"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testResponseCode401() { // 401 = unauthorized
        let xpectation = expectation(description: "401")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 401))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("invalid api key"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testResponseCode500() { // 500 = internal server error
        let xpectation = expectation(description: "500")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 500))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("internal server error"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testNon200ResponseCode() { // 302 = redirection
        let xpectation = expectation(description: "non 200")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: MockNetworkSession(statusCode: 302))
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("non-200 response"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testNoNetworkResponse() {
        let xpectation = expectation(description: "no network response")
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        createApiClient(networkSession: NoNetworkNetworkSession())
            .send(iterableRequest: iterableRequest).onError { sendError in
                xpectation.fulfill()
                XCTAssert(sendError.reason!.lowercased().contains("nsurlerrordomain"))
            }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    func testNetworkTimeoutResponse() {
        let xpectation = expectation(description: "timeout network response")
        
        let responseTime = 2.0
        let timeout = 0.1
        
        HTTPStubs.stubRequests(passingTest: { (_) -> Bool in
            true
        }) { (_) -> HTTPStubsResponse in
            let response = HTTPStubsResponse(data: try! JSONSerialization.data(withJSONObject: [:], options: []),
                                             statusCode: 200,
                                             headers: nil)
            response.requestTime = 0.0
            response.responseTime = responseTime
            return response
        }
        
        let networkSession = URLSession(configuration: URLSessionConfiguration.default)
        
        let iterableRequest = IterableRequest.post(PostRequest(path: "", args: nil, body: [:]))
        
        let apiClient = createApiClient(networkSession: networkSession)
        var urlRequest = apiClient.convertToURLRequest(iterableRequest: iterableRequest)!
        urlRequest.timeoutInterval = timeout
        
        NetworkHelper.sendRequest(urlRequest, usingSession: networkSession).onError { sendError in
            xpectation.fulfill()
            XCTAssert(sendError.reason!.lowercased().contains("timed out"))
        }
        
        wait(for: [xpectation], timeout: testExpectationTimeout)
    }
    
    private func verifyIterableHeaders(_ urlRequest: URLRequest) {
        XCTAssertEqual(urlRequest.value(forHTTPHeaderField: JsonKey.Header.sdkPlatform), JsonValue.iOS.jsonStringValue)
        XCTAssertEqual(urlRequest.value(forHTTPHeaderField: JsonKey.Header.sdkVersion), IterableAPI.sdkVersion)
        XCTAssertEqual(urlRequest.value(forHTTPHeaderField: JsonKey.Header.apiKey), apiKey)
        XCTAssertEqual(urlRequest.value(forHTTPHeaderField: JsonKey.contentType.jsonKey), JsonValue.applicationJson.jsonStringValue)
    }
    
    private func verifyAuthTokenInHeader(_ urlRequest: URLRequest, _ authToken: String) {
        XCTAssertEqual(urlRequest.value(forHTTPHeaderField: JsonKey.Header.authorization), "\(JsonValue.bearer.rawValue) \(authToken)")
    }
    
    private func createApiClient(networkSession: NetworkSessionProtocol = MockNetworkSession()) -> ApiClient {
        ApiClient(apiKey: apiKey,
                  authProvider: AuthProviderNoToken(),
                  endPoint: Endpoint.api,
                  networkSession: networkSession,
                  deviceMetadata: IterableAPIInternal.initializeForTesting().deviceMetadata)
    }
    
    private func createApiClientWithAuthToken() -> ApiClient {
        ApiClient(apiKey: apiKey,
                  authProvider: self,
                  endPoint: Endpoint.api,
                  networkSession: MockNetworkSession(),
                  deviceMetadata: IterableAPIInternal.initializeForTesting().deviceMetadata)
    }
}

extension IterableAPIResponseTests: AuthProvider {
    var auth: Auth {
        Auth(userId: nil, email: email, authToken: authToken)
    }
}

class AuthProviderNoToken: AuthProvider {
    var auth: Auth {
        Auth(userId: nil, email: "user@example.com", authToken: nil)
    }
}
